text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1019 @@
+  proposed in a separate RFC, parallel to this one.
+
+## Motivation
+
+Glimmer components are a simpler base component class that encourage template
+oriented design and thinner class definitions. They are easier to understand in
+isolation, and require less knowledge of the framework to use effectively.
+
+This example shows a component written with the classic component API:
+
+```hbs
+<!-- templates/components/post.hbs -->
+{{#if (eq this.type 'image'}}
+  <img src={{@post.imageUrl}} title={{@post.imageTitle}}/>
+{{/if}}
+
+{{@post.text}}
+```
+```js
+// components/post.js
+export default Component.extend({
+  tagName: 'section',
+  classNames: ['post'],
+  classNameBindings: ['type'],
+  ariaRole: 'region',
+
+  /* Arguments */
+  post: null,
+
+  type: readOnly('post.type'),
+
+  didInsertElement() {
+    if (this.type === 'image') {
+      setupImageOverlay(this.element.querySelector('img'));
+    }
+  }
+});
+```
+
+And here is an equivalent component written with the Glimmer component API:
+
+```hbs
+<!-- templates/components/post.hbs -->
+<section ...attributes role="region" type={{@post.type}} class="post {{@post.type}}">
+  {{#if (eq @post.type 'image')}}
+    <img {{did-render this.didInsertImage}} src={{@post.imageUrl}} title={{@post.imageTitle}} />
+  {{/if}}
+
+  {{@post.text}}
+</section>
+```
+```js
+// components/post.js
+export default class PostComponent extends GlimmerComponent {
+  didInsertImage(element) {
+    setupImageOverlay(element);
+  }
+}
+```
+
+Glimmer components eliminate many of the common paper cuts that cause confusion
+with classic components, and align more closely with modern template syntax and
+features.
+
+### Outer HTML Semantics
+
+The biggest change Glimmer components make is defaulting to outer HTML
+semantics. In the classic component API, components had a implicit wrapper
+element. Given this component template:
+
+```hbs
+Hello, world!
+```
+
+The output by _default_ would be something like:
+
+```html
+<!-- OUTPUT -->
+<div id="ember-1234" class="ember-view">
+  Hello, world!
+</div>
+```
+
+But we can't know that for sure unless we look at the component definition. If
+we do, we might see that the outer wrapping element is actually a `section`, and
+it has a `.hello-world` class:
+
+```js
+export default Component.extend({
+  tagName: 'section',
+  classNames: ['hello-world']
+});
+```
+```html
+<!-- OUTPUT -->
+<section id="ember-1234" class="hello-world ember-view">
+  Hello, world!
+</section>
+```
+
+This behavior means that the template for a component is missing crucial
+information and context. Even for the simplest component, users must check the
+class definition to know with certainty what the full template of the component
+is. And unlike bindings, there is no hint to the user that there may be
+something dynamic that they should check on - without advanced knowledge of
+Ember's APIs, there is no way of knowing about this behavior.
```

Comment:
"tagless" (`tagName: ''`) components deserve a place in this comparative analysis imo